New CGEN error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 e476a13, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 45fd7eb.e476a13
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |24 cpus, 64bit, little endian, AMD Ryzen 9 5900X 12-Core Processor  
|Memory              |73.76GB/127.92GB
|                    |
|V executable        |S:\repo\vlang\v.exe
|V last modified time|2025-01-26 19:34:50
|                    |
|V home dir          |OK, value: S:\repo\vlang
|VMODULES            |OK, value: C:\Users\john3\.vmodules
|VTMP                |OK, value: C:\Users\john3\AppData\Local\Temp\v_0
|Current working dir |OK, value: S:\vProjects\vSelenium
|                    |
|Git version         |git version 2.45.1.windows.1
|V git status        |weekly.2025.03-89-ge476a130
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |gcc (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders, r2) 14.2.0
|clang version       |(built by Brecht Sanders, r2) clang version 19.1.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg -cg main.v && S:\vProjects\vSelenium\main.exe`
See attached file `main.v`

**What did you see?**
```
main.v:4:8: warning: module 'common (webdriver.common)' is imported but never used
    2 | 
    3 | import webdriver.edge { EdgeDriver, EdgeOptions, EdgeService }
    4 | import webdriver.common
      |        ~~~~~~~~~~~~~~~~
    5 | 
    6 | fn open_browser(download_path string) ?EdgeDriver[EdgeOptions, EdgeService] {
webdriver/chromium/web_driver.v:4:8: warning: module 'common (webdriver.common)' is imported but never used
    2 | 
    3 | import webdriver.remote { RemoteWebDriver }
    4 | import webdriver.common { DriverFinder }
      |        ~~~~~~~~~~~~~~~~
    5 | 
    6 | // ChromiumDriver - Create a new WebDriver instance of the ChromiumDriver. Starts the
webdriver/common/selenium_manager.v:140:3: warning: unused variable: `arch`
  138 |         s.machine = un.machine
  139 |         s.user_os = os.user_os()
  140 |         arch := if s.user_os in ['linux', 'freebsd', 'openbsd'] { s.machine } else { 'any' }
      |         ~~~~
  141 |         if s.user_os in ['freebsd', 'openbsd'] {
  142 |             log.warn('Selenium Manager binary may not be compatible with ${s.user_os}; verify settings')
C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:860: warning: WINVER redefined

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:1262: warning: InterlockedIncrement64 redefined

In file included from C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:1642:

s:/repo/vlang/thirdparty/tcc/include/winapi/synchapi.h:13: warning: CreateEvent redefined

In file included from C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:1642:

s:/repo/vlang/thirdparty/tcc/include/winapi/synchapi.h:117: warning: OpenEvent redefined

In file included from C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:1642:

s:/repo/vlang/thirdparty/tcc/include/winapi/synchapi.h:167: warning: CreateMutex redefined

In file included from C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:1809:

S:/repo/vlang/thirdparty/vschannel/vschannel.c:284: warning: implicit declaration of function 'WSAConnectByNameW'

In file included from C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:1809:

S:/repo/vlang/thirdparty/vschannel/vschannel.c:300: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:12901: warning: implicit declaration of function 'tcc_backtrace'

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:20281: warning: implicit declaration of function 'GetFinalPathNameByHandleW'

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:26900: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:26913: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:26930: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:26937: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:26975: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:26985: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:27499: warning: implicit declaration of function 'inet_ntop'

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:27499: warning: cast between pointer and integer of different size

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:27509: warning: cast between pointer and integer of different size

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:28238: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:32354: warning: assignment from incompatible pointer type

C:/Users/john3/AppData/Local/Temp/v_0/main.01JJJ1SCWYDMTWJVZC6KDZVN4A.tmp.c:32591: error: pointer expected
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

clean compile



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -4312,7 +4312,9 @@ fn (mut c Checker) smartcast(mut expr ast.Expr, cur_type ast.Type, to_type_ ast.
 			mut orig_type := 0
 			mut is_inherited := false
 			mut ct_type_var := ast.ComptimeVarKind.no_comptime
+			mut is_ct_type_unwrapped := false
 			if mut expr.obj is ast.Var {
+				is_ct_type_unwrapped = expr.obj.ct_type_var != ast.ComptimeVarKind.no_comptime
 				is_mut = expr.obj.is_mut
 				smartcasts << expr.obj.smartcasts
 				is_already_casted = expr.obj.pos.pos == expr.pos.pos
@@ -4334,16 +4336,17 @@ fn (mut c Checker) smartcast(mut expr ast.Expr, cur_type ast.Type, to_type_ ast.
 						if cur_type.has_flag(.option) && !to_type.has_flag(.option) {
 							if !var.is_unwrapped {
 								scope.register(ast.Var{
-									name:         expr.name
-									typ:          cur_type
-									pos:          expr.pos
-									is_used:      true
-									is_mut:       expr.is_mut
-									is_inherited: is_inherited
-									smartcasts:   [to_type]
-									orig_type:    orig_type
-									ct_type_var:  ct_type_var
-									is_unwrapped: true
+									name:              expr.name
+									typ:               cur_type
+									pos:               expr.pos
+									is_used:           true
+									is_mut:            expr.is_mut
+									is_inherited:      is_inherited
+									smartcasts:        [to_type]
+									orig_type:         orig_type
+									ct_type_var:       ct_type_var
+									ct_type_unwrapped: is_ct_type_unwrapped
+									is_unwrapped:      true
 								})
 							} else {
 								scope.update_smartcasts(expr.name, to_type, true)
@@ -4355,16 +4358,17 @@ fn (mut c Checker) smartcast(mut expr ast.Expr, cur_type ast.Type, to_type_ ast.
 					}
 				}
 				scope.register(ast.Var{
-					name:         expr.name
-					typ:          cur_type
-					pos:          expr.pos
-					is_used:      true
-					is_mut:       expr.is_mut
-					is_inherited: is_inherited
-					is_unwrapped: is_option_unwrap
-					smartcasts:   smartcasts
-					orig_type:    orig_type
-					ct_type_var:  ct_type_var
+					name:              expr.name
+					typ:               cur_type
+					pos:               expr.pos
+					is_used:           true
+					is_mut:            expr.is_mut
+					is_inherited:      is_inherited
+					is_unwrapped:      is_option_unwrap
+					smartcasts:        smartcasts
+					orig_type:         orig_type
+					ct_type_var:       ct_type_var
+					ct_type_unwrapped: is_ct_type_unwrapped
 				})
 			} else if is_mut && !expr.is_mut {
 				c.smartcast_mut_pos = expr.pos

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5158,7 +5158,8 @@ fn (mut g Gen) ident(node ast.Ident) {
 							}
 						}
 						if i == 0 && node.obj.ct_type_var != .smartcast && node.obj.is_unwrapped {
-							dot := if !node.obj.orig_type.is_ptr() && obj_sym.is_heap() {
+							dot := if !node.obj.ct_type_unwrapped && !node.obj.orig_type.is_ptr()
+								&& obj_sym.is_heap() {
 								'->'
 							} else {
 								'.'

--- a/vlib/v/tests/comptime/comptime_var_unwrap_test.v
+++ b/vlib/v/tests/comptime/comptime_var_unwrap_test.v
@@ -1,0 +1,26 @@
+struct Foo {
+	a int
+}
+
+fn Foo.init[T](a T) {
+	dump(a)
+}
+
+@[heap]
+struct Bar {
+	field int
+}
+
+fn t[T](a ?T) T {
+	mut b := a
+	if b != none {
+		Foo.init(b)
+		return b
+	}
+	assert false
+	return T{}
+}
+
+fn test_main() {
+	assert t(Bar{}) == Bar{}
+}

